src/EventStore.Core.Tests/DataStructures/bloom_filter_integrity_should.cs

```diff
@@ -0,0 +1,125 @@
+﻿using System;
+using System.Linq;
+using System.Runtime.InteropServices;
+using EventStore.Core.DataStructures.ProbabilisticFilter.MemoryMappedFileBloomFilter;
+using EventStore.Core.Index.Hashes;
+using NUnit.Framework;
+
+namespace EventStore.Core.Tests.DataStructures {
+	[TestFixture]
+	public class bloom_filter_integrity_should {
+		private static byte[] NewCacheLine(byte b) {
+			var xs = new byte[64];
+			for (int i = 0; i < xs.Length; i++) {
+				xs[i] = b;
+			}
+			return xs;
+		}
+
+		[DatapointSource]
+		public static object[][] RecoverCases = new object[][] {
+			new object[] { "all zeros is valid", true, NewCacheLine(0x00) },
+			new object[] { "all zeros with hash is valid", true, NewCacheLine(0x00).WithHash() },
+			new object[] { "all ones is invalid", false, NewCacheLine(0xFF) },
+			new object[] { "all ones with hash is valid", true, NewCacheLine(0xFF).WithHash() },
+			new object[] { "data but no hash is invalid", false, NewCacheLine(0x00).With(3, 0xAB) },
+			new object[] { "data with hash is valid", true, NewCacheLine(0x00).With(3, 0xAB).WithHash() },
+		};
+
+		[TestCaseSource(nameof(RecoverCases))]
+		public void recover_correctly(string name, bool expected, byte[] bytes) {
+			var originalBytes = bytes.ToArray();
+
+			var byteSpan = bytes.AsSpan();
+
+			if (expected) {
+				Assert.True(BloomFilterIntegrity.ValidateHash(byteSpan));
+
+				// bytes should be unchanged
+				for (var i = 0; i < bytes.Length; i++) {
+					Assert.AreEqual(originalBytes[i], bytes[i]);
+				}
+			} else {
+				Assert.False(BloomFilterIntegrity.ValidateHash(byteSpan));
+
+				// bytes should be all 1s
+				foreach (var b in bytes) {
+					Assert.AreEqual(0b1111_1111, b);
+				}
+			}
+		}
+
+		[DatapointSource]
+		public static object[][] WriteCases = new object[][] {
+			new object[] { "all 0x00", true, NewCacheLine(0x00) },
+			new object[] { "all 0xAA", true, NewCacheLine(0xAA) },
+			new object[] { "all 0xFF - corrupted and made safe", false, NewCacheLine(0xFF) },
+		};
+
+		[TestCaseSource(nameof(WriteCases))]
+		public void write_correctly(string name, bool expected, byte[] bytes) {
+			var originalBytes = bytes.ToArray();
+			var byteSpan = bytes.AsSpan();
+
+			BloomFilterIntegrity.WriteHash(byteSpan);
+
+			// bytes before the hash should be unchanged
+			for (var i = 0; i < bytes.Length - sizeof(uint); i++) {
+				Assert.AreEqual(originalBytes[i], bytes[i]);
+			}
+
+			// cacheline should be valid
+			if (expected) {
+				Assert.True(BloomFilterIntegrity.ValidateHash(byteSpan));
+			} else {
+				Assert.False(BloomFilterIntegrity.ValidateHash(byteSpan));
+			}
+		}
+
+		[TestCase]
+		public void rehash_when_one_too_many_bits_are_set() {
```

Comment:
```suggestion
		public void rehash_when_too_many_one_bits_are_set() {
```

Reply:
it's supposed to describe the case where there is one too many bits set, as in only one extra

Reply:
maybe "rehash_when_exactly_one_too_many_bits_are_set" ?